samples/msal-node-samples/on-behalf-of-distributed-cache/src/AuthProvider.ts

```diff
@@ -0,0 +1,154 @@
+        this.msalConfig = msalConfig;
+        this.cryptoProvider = new CryptoProvider();
+        this.cacheClientWrapper = new RedisClientWrapper(cacheClient);
+    }
+
+    static async initialize(appConfig: AppConfig, cacheClient: RedisClientType): Promise<AuthProvider> {
+        const msalConfig = {
+            auth: {
+                clientId: appConfig.clientId,
+                authority: `${appConfig.instance}/${appConfig.tenantId}`,
+                clientSecret: appConfig.clientSecret,
+            },
+            system: {
+                loggerOptions: {
+                    loggerCallback(loglevel, message, containsPii) {
+                        console.log(message);
+                    },
+                    logLevel: LogLevel.Trace,
+                    piiLoggingEnabled: false
+                },
+                // proxyUrl: "http://localhost:8888" // uncomment to capture traffic with Fiddler
+            }
+        } as Configuration;
+
+        const msalConfigWithMetadata = await AuthProvider.getMetadata(msalConfig, cacheClient);
+
+        return new AuthProvider(msalConfigWithMetadata, cacheClient);
+    }
+
+    async getToken(tokenRequest: OnBehalfOfRequest): Promise<AuthenticationResult | null> {
+        const partitionKey = await this.cryptoProvider.hashString(tokenRequest.oboAssertion);
+
+        const cca = new ConfidentialClientApplication({
+            ...this.msalConfig,
+            cache: {
+                cachePlugin: new CustomCachePlugin(
+                    this.cacheClientWrapper,
+                    partitionKey // partitionKey hash(oboAssertion)
+                )
+            }
+        });
+
+        let tokenResponse = null;
+
+        try {
+            performance.mark("acquireTokenOnBehalfOf-start");
+            await cca.getTokenCache().getAllAccounts(); // required for cache read
+            tokenResponse = await cca.acquireTokenOnBehalfOf(tokenRequest);
+            performance.mark("acquireTokenOnBehalfOf-end");
+            performance.measure(
+                tokenResponse?.fromCache ? "acquireTokenOnBehalfOf-fromCache" : "acquireTokenOnBehalfOf-fromNetwork",
+                "acquireTokenOnBehalfOf-start",
+                "acquireTokenOnBehalfOf-end"
+            );
+        } catch (error) {
+            throw error;
+        }
+
+        return tokenResponse;
+    }
+
+    private static async getMetadata(msalConfig: Configuration, cacheClient: RedisClientType): Promise<Configuration> {
+        const msalConfigWithMetadata = msalConfig;
+
+        const clientId = msalConfigWithMetadata.auth.clientId;
+        const tenantId = msalConfigWithMetadata.auth.authority!.split("/").pop()!;
+
+        try {
+            let [cloudDiscoveryMetadata, authorityMetadata] = await Promise.all([
+                cacheClient.get(`${clientId}.${tenantId}.discovery-metadata`),
+                cacheClient.get(`${clientId}.${tenantId}.authority-metadata`)
```

Comment:
Interesting, so you're caching the metadata in Redis. 

1. I would add a potential improvement if possible - set the TTL for these entries, I believe the AAD endpoints suggest how long to cache this for (24h?). 
2. the key should be `{authority}.discovery-metadata` ? It's not tied to clientid, but it is tied to the cloud. `{tenand_id}.discoevry_metadata` will probably work as well due to GUIDs being unique and all.

Reply:
1. sure thing, I think we can do this
2. I see. if tenantId alone is ok, would be fine to leave it as it is i.e. clientId.tenantId (asking for consisentency, otherwise I have to go back and update the other samples :D )

Reply:
For #2 yes, it's fine. Redis will not be shared cross-cloud anyway.